### PR TITLE
Improve consistency of cache key generation with args/kwargs

### PR DESCRIPTION
**Pull request by nbecker-yc**
### Overview
Ticket: https://app.asana.com/1/1144271657183267/project/1203633948717779/task/1209820190967177?focus=true

### How to test
- [x] Review code/run tests

### CODE REVIEW

**Review by dlim-yc — COMMENTED** [review]
Started by reviewing the tests, and came up with this edge case where the cache key generation could be optimized even further:

```
        inc_1 = incrementer.instance_increment_by_with_kwargs(1, None)
        inc_2 = incrementer.instance_increment_by_with_kwargs(1, useless_kwarg=None)
        inc_3 = incrementer.instance_increment_by_with_kwargs(1)
```

All three of these could technically be cached at the same key, since the default value of the kwarg is `None`. I believe you covered for 1 == 2, but not 3

Does this seem feasible?

**Comment by nbecker-yc** [comment]
> Started by reviewing the tests, and came up with this edge case where the cache key generation could be optimized even further:
> 
> ```
>         inc_1 = incrementer.instance_increment_by_with_kwargs(1, None)
>         inc_2 = incrementer.instance_increment_by_with_kwargs(1, useless_kwarg=None)
>         inc_3 = incrementer.instance_increment_by_with_kwargs(1)
> ```
> 
> All three of these could technically be cached at the same key, since the default value of the kwarg is `None`. I believe you covered for 1 == 2, but not 3
> 
> Does this seem feasible?

@dlim-yc great catch. just fixed this and added 3 additional tests (one for each type of decorator).

**Review by dlim-yc — APPROVED** [review]
Looks great! Pretty elegant solution

File: cache_helper/decorators.py [hunk 1]
```diff
@@ -15,14 +17,32 @@
...
 def cached(timeout):
     def _cached(func):
         func_name = utils.get_function_name(func)
+        func_signature = signature(func)
 
         @wraps(func)
         def wrapper(*args, **kwargs):
-            function_cache_key = utils.get_function_cache_key(func_name, args, kwargs)
-            cache_key = utils.get_hashed_cache_key(function_cache_key)
+            cache_key, function_cache_key = _get_function_cache_keys(func_name, func_signature, args, kwargs)
```
[the remark this answers is not in the document]
Comment (nbecker-yc) [reply cache_helper/decorators.py:45]: yeah definitely agreed. I left the names as they were to minimize changes but agree that's cleaner. I'll update accordingly.

[the remark this answers is not in the document]
Comment (nbecker-yc) [reply cache_helper/decorators.py]: done!
